migrations/repotracker_cleanup.go

```diff
@@ -0,0 +1,452 @@
+				}
+
+			case buildStatusesKey:
+				if err := rawD[i].Value.Unmarshal(&bstatuses); err != nil {
+					return errors.Wrap(err, "error unmarshaling build statuses")
+				}
+
+			case buildIDsKey:
+				if err := rawD[i].Value.Unmarshal(&builds); err != nil {
+					return errors.Wrap(err, "error unmarshaling builds")
+				}
+			}
+		}
+
+		buildMap := map[string]bool{}
+
+		newBuilds := []string{}
+		for _, buildID := range builds {
+			_, ok := buildMap[buildID]
+			if !ok {
+				query := session.DB(database).C(buildCollection).FindId(buildID)
+				err := query.One(nil)
+				if err == mgo.ErrNotFound {
+					continue
+				}
+				if err != nil {
+					return errors.Wrapf(err, "error fetching build %s", buildID)
+				}
+				buildMap[buildID] = true
+			}
+
+			newBuilds = append(newBuilds, buildID)
+		}
+
+		newBuildStatuses := []version.BuildStatus{}
+		for _, bstatus := range bstatuses {
+			_, ok := buildMap[bstatus.BuildId]
+			if !ok {
+				query := session.DB(database).C(buildCollection).FindId(bstatus.BuildId)
+				err := query.One(nil)
+				if err == mgo.ErrNotFound {
+					continue
+				}
+				if err != nil {
+					return errors.Wrapf(err, "error fetching build %s", bstatus.BuildId)
+				}
+				buildMap[bstatus.BuildId] = true
+			}
+
+			newBuildStatuses = append(newBuildStatuses, bstatus)
+		}
+
+		if len(newBuilds) == 0 && len(newBuildStatuses) == 0 {
+			return session.DB(database).C(versionCollection).RemoveId(versionID)
+		}
+
+		update := bson.M{
+			"$set": bson.M{
+				buildIDsKey:      newBuilds,
+				buildStatusesKey: newBuildStatuses,
+			},
+		}
+
+		return session.DB(database).C(versionCollection).UpdateId(versionID, update)
+	}
+}
+
+func orphanedBuildCleanupGenerator(env anser.Environment, db string, limit int) (anser.Generator, error) {
+	const migrationName = "clean_orphaned_builds"
+
+	if err := env.RegisterManualMigrationOperation(migrationName, makeOrphanedBuildCleanup(db)); err != nil {
+		return nil, err
+	}
+
+	opts := model.GeneratorOptions{
+		NS: model.Namespace{
+			DB:         db,
+			Collection: buildCollection,
+		},
+		Query: bson.M{
+			requesterKey: evergreen.RepotrackerVersionRequester,
+			statusKey:    evergreen.BuildCreated, // TODO is this true?
+		},
+		Limit: limit,
+		JobID: "migration-builds-orphans",
+	}
+
+	return anser.NewManualMigrationGenerator(env, opts, migrationName), nil
+}
+
+// orphanedBuildCleanup, given a build, will:
+// 1. If it's version doesn't exist, remove the build and it's tasks
+// 2. Check it's task cache, and ensure that all listed tasks exist, removing
+//    any that do not exist
+func makeOrphanedBuildCleanup(database string) db.MigrationOperation {
+	const (
+		idKey      = "_id"
+		tasksKey   = "tasks"
+		versionKey = "version"
+		buildIDKey = "build_id"
+	)
+	return func(session db.Session, rawD bson.RawD) error {
+		defer session.Close()
+
+		cachedTasks := []build.TaskCache{}
+		versionID := ""
+		buildID := ""
+
+		for i := range rawD {
```

Comment:
This looks pretty similar to https://github.com/evergreen-ci/evergreen/pull/647/files#diff-ca6f59eb4ee18cb99323fc99c38e29d4R62